### DISCUSSION TITLE
Parametrize a test with a list from a fixture?

### DISCUSSION THREAD

**Discussion by ghost** [pytest-dev/pytest#10571] [Q&A] [answered]
Here's my fixture:
```python
@fixture
def devices(scope='module');
    with connect(get_all_devices()) as connected_devices:
        yield connected_devices
```
Here's how I want to use it:
```python
@parametrize('device', devices)
def test_routine(device):
    device.do_something()
    assert device.is_ok()
```
What I'm doing instead:
```python
def test_routine(subtests, devices):
    for index, device in enumerate(devices):
        with subtests.test(device, i=index):
            device.do_something()
            assert device.is_ok()
```
Is there any way to do this?

**Comment by RonnyPfannschmidt** [comment, marked as answer]
no, fixtures happen after parametrize

currently there is no plan/effort to enable a parametrize scope for them

**Comment by The-Compiler** [comment]
You can probably get to a similar result with a custom marker and [`pytest_generate_tests` hook](https://docs.pytest.org/en/stable/how-to/parametrize.html#pytest-generate-tests) instead.

**Reply by ghost** [reply]
Can you actually get the fixture results in the hook?

**Reply by RonnyPfannschmidt** [reply]
no, fixtures happen after collection, generate_tests happens at collection

**Reply by ghost** [reply]
> no, fixtures happen after collection, generate_tests happens at collection

thanks, that's how I understood it as well

**Comment by BorjaEst** [comment]
This is how I would do:
````py
@fixture(scope='module', params=devices)
def device(request):
    with connect(request.param) as device:
        yield connected_device

def test_routine(device):
    device.do_something()
    assert device.is_ok()`
````

Note this will fail if the way you collect `devices` raises an error, but there is a difference between the meaning of test fail and error, test fail is when your asserts are incorrect.

If you do not want the testing process to crash during the collection process or prefer some more explicit names for each of the fixture parametrization, you should resolve/test the driver on the fixture runtime:

````py
@fixture(scope='module', params=["driver_string1", "driver_string2"])
def device(request):
    device_driver = driver_from_string(request.param)
    with connect(request.param) as device:
        device.verify_something_before_test()
        yield connected_device
````

Here I add also an example from my packages:
https://github.com/BorjaEst/gevopy/blob/main/tests/conftest.py#L31

**Reply by ghost** [reply]
Thanks for the suggestion, though I think it does not work with the system I'm using as all the devices need to be connected at the same time for the tests to work, as in the fixture should only run once and return a list.
The way I had it with unittest before is I used the setUpClass and tearDownClass to (dis)connect everything and iterate through the devices with subtests, same as in my example here, however it would be much much nicer if it could work with parameterize. I find the pytest subtests a bit uncomfortable to use.
